src/crimson/osd/pg.cc

```diff
@@ -851,4 +859,470 @@ void PG::handle_rep_op_reply(crimson::net::Connection* conn,
+  for (auto p : replicas_by_num_missing) {
+    ret.push_back(p.second);
+  }
+  return ret;
+}
+
+size_t PG::start_replica_recovery_ops(
+  size_t max_to_start,
+  std::vector<crimson::osd::blocking_future<>> *out)
+{
+  if (!is_recovering()) {
+    return 0;
+  }
+  uint64_t started = 0;
+
+  ceph_assert(!peering_state.get_acting_recovery_backfill().empty());
+
+  auto recovery_order = get_replica_recovery_order();
+  for (auto &peer: recovery_order) {
+    ceph_assert(peer != peering_state.get_primary());
+    auto pm = peering_state.get_peer_missing().find(peer);
+    ceph_assert(pm != peering_state.get_peer_missing().end());
+
+    size_t m_sz = pm->second.num_missing();
+
+    logger().debug(
+      "{}: peer osd.{} missing {} objects",
+      __func__,
+      peer,
+      m_sz);
+    logger().trace(
+      "{}: peer osd.{} missing {}",
+      __func__,
+      peer,
+      pm->second.get_items());
+
+    // recover oldest first
+    const pg_missing_t &m(pm->second);
+    for (auto p = m.get_rmissing().begin();
+	 p != m.get_rmissing().end() && started < max_to_start;
+	 ++p) {
+      const auto &soid = p->second;
+
+      if (peering_state.get_missing_loc().is_unfound(soid)) {
+	logger().debug(
+	  "{}: object {} still unfound",
+	  __func__,
+	  soid);
+	continue;
+      }
+
+      const pg_info_t &pi = peering_state.get_peer_info(peer);
+      if (soid > pi.last_backfill) {
+	if (!backend->is_recovering(soid)) {
+	  logger().error(
+	    "{}: object {} in missing set for backfill (last_backfill {})"
+	    " but not in recovering",
+	    __func__,
+	    soid,
+	    pi.last_backfill);
+	  ceph_abort();
+	}
+	continue;
+      }
+
+      if (backend->is_recovering(soid)) {
+	logger().debug(
+	  "{}: already recovering object {}",
+	  __func__,
+	  soid);
+	continue;
+      }
+
+      if (peering_state.get_missing_loc().is_deleted(soid)) {
+	logger().debug(
+	  "{}: soid {} is a delete, removing",
+	  __func__,
+	  soid);
+	map<hobject_t,pg_missing_item>::const_iterator r =
+	  m.get_items().find(soid);
+	started += prep_object_replica_deletes(
+	  soid, r->second.need, out);
+	continue;
+      }
+
+      if (soid.is_snap() &&
+	  peering_state.get_pg_log().get_missing().is_missing(
+	    soid.get_head())) {
+	logger().debug(
+	  "{}: head {} still missing on primary",
+	  __func__,
+	  soid.get_head());
+	continue;
+      }
+
+      if (peering_state.get_pg_log().get_missing().is_missing(soid)) {
+	logger().debug(
+	  "{}: soid {} still missing on primary",
+	  __func__,
+	  soid);
+	continue;
+      }
+
+      logger().debug(
+	"{}: recover_object_replicas({})",
+	__func__,
+	soid);
+      map<hobject_t,pg_missing_item>::const_iterator r = m.get_items().find(
+	soid);
+      started += prep_object_replica_pushes(
+	soid, r->second.need, out);
+    }
+  }
+
+  return started;
+}
+
+size_t PG::start_backfill_ops(
+  size_t max_to_start,
+  std::vector<crimson::osd::blocking_future<>> *out)
+{
+  if (!is_backfilling()) {
+    return 0;
+  }
```

Comment:
why do we need this check?

Reply:
Oh, this is part of backfill mechanism which is not implemented here yet, I'll replace it with an "not implemented" abort